lib/stripe.ex

```diff
@@ -1,149 +1,266 @@
+    config :stripity_stripe, api_key: YOUR_SECRET_KEY
+    """
+  end
 
-  defmodule MissingSecretKeyError do
+  defmodule APIAuthenticationError do
     defexception message: """
-      The secret_key setting is required so that we can report the
-      correct environment instance to Stripe. Please configure
-      secret_key in your config.exs and environment specific config files
-      to have accurate reporting of errors.
-      config :stripity_stripe, secret_key: YOUR_SECRET_KEY
+    Stripe could not authenticate the request with the provided API key
     """
   end
 
+  defmodule APIRateLimitingError do
+    defexception message: """
+    Stripe is currently limiting the rate at which it accepts your HTTP requests
+    """
+  end
 
-  @doc """
-  Grabs STRIPE_SECRET_KEY from system ENV
-  Returns binary
-  """
-  def config_or_env_key do
-    require_stripe_key()
+  defmodule APIError do
+    defexception [:message, :type, :status_code, :code]
+
+    @type t :: %__MODULE__{
+      message: String.t | nil,
+      type: String.t | nil,
+      status_code: pos_integer | nil,
+      code: String.t | nil
+    }
+
+    @spec exception({integer, map}) :: t
+    def exception({status_code, %{"type" => type, "message" => message} = body}) do
+      # code is not a guaranteed key
+      code = Map.get(body, "code")
+
+      %__MODULE__{
+        message: message,
+        type: type,
+        status_code: status_code,
+        code: code
+      }
+    end
+
+    def exception({status_code, _}) do
+      msg = """
+      The Stripe HTTP client received an error response with status code #{status_code}
+      """
+
+      %__MODULE__{
+        message: msg,
+        status_code: status_code
+      }
+    end
   end
 
-  @doc """
-  Grabs STRIPE_PLATFORM_CLIENT_ID from system ENV
-  Returns binary
-  """
-  def config_or_env_platform_client_id do
-    Application.get_env(:stripity_stripe, :platform_client_id) || System.get_env "STRIPE_PLATFORM_CLIENT_ID"
+  defmodule HTTPError do
+    defexception message: """
+    The Stripe HTTP client encountered an error while communicating with the
+    Stripe service.
+    """
   end
 
+  alias __MODULE__.{MissingAPIKeyError, HTTPError, APIRateLimitingError, APIError}
+
+  @type method :: :get | :post | :put | :delete | :patch
+  @type headers :: %{required(String.t) => String.t}
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @pool_name __MODULE__
+
   @doc """
-  Creates the URL for our endpoint.
-  Args:
-    * endpoint - part of the API we're hitting
-  Returns string
+  Starts necessary components for the HTTP client to function
   """
-  def process_url(endpoint) do
-    "https://api.stripe.com/v1/" <> endpoint
+  @spec start() :: :ok
+  def start() do
+    if use_pool?() do
+      pool_options = get_pool_options()
+      :ok = :hackney_pool.start_pool(@pool_name, pool_options)
+    end
+
+    :ok
   end
```

Comment:
@schneidmaster given your issues you were talking about with pools in Hackney, do you see any issues here?